Mathlib/Analysis/Calculus/ContDiff/FaaDiBruno.lean

```diff
@@ -0,0 +1,694 @@
+map of the `vᵢ`. For this, instead of working with partitions of `{0, ..., n-1}` and ordering their
+parts, we work with partitions in which the ordering is part of the data -- this is equivalent,
+but much more convenient to implement. We call these `OrderedFinpartition n`.
+
+Note that the implementation of `OrderedFinpartition` is very specific to the Faa di Bruno formula:
+as testified by the formula above, what matters is really the embedding of the parts in `Fin n`,
+and moreover the parts have to be ordered by `max I₀ < ... < max Iₖ₋₁` for the formula to hold
+in the general case where the iterated differential might not be symmetric. The defeqs with respect
+to `Fin.cons` are also important when doing the induction. For this reason, we do not expect this
+class to be useful beyond the Faa di Bruno formula, which is why it is in this file instead
+of a dedicated file in the `Combinatorics` folder.
+
+## Main results
+
+Given `c : OrderedFinpartition n` and two formal multilinear series `q` and `p`, we
+define `c.compAlongOrderedFinpartition q p` as an `n`-multilinear map given by the formula above,
+i.e., `(v₁, ..., vₙ) ↦ qₖ (p_{i₁} (v_{I₁}), ..., p_{iₖ} (v_{Iₖ}))`.
+
+Then, we define `q.taylorComp p` as a formal multilinear series whose `n`-th term is
+the sum of `c.compAlongOrderedFinpartition q p` over all ordered finpartitions of size `n`.
+
+Finally, we prove in `HasFTaylorSeriesUptoOn.comp` that, if two functions `g` and `f` have Taylor
+series up to `n` given by `q` and `p`, then `g ∘ f` also has a Taylor series,
+given by `q.taylorComp p`.
+
+## Implementation
+
+A first technical difficulty is to implement the extension process of `OrderedFinpartition`
+corresponding to adding a new atom, or appending an atom to an existing part, and defining the
+associated increasing parameterizations that show up in the definition
+of `compAlongOrderedFinpartition`.
+
+Then, one has to show that the ordered finpartitions thus
+obtained give exactly all ordered finpartitions of order `n+1`. For this, we define the inverse
+process (shrinking a finpartition of `n+1` by erasing `0`, either as an atom or from the part
+that contains it), and we show that these processes are inverse to each other, yielding an
+equivalence between `((c : OrderedFinpartition n) × Option (Fin c.length))`
+and `OrderedFinpartition (n + 1)`. This equivalence shows up prominently in the inductive proof
+of Faa di Bruno formula to identify the sums that show up.
+
+## Current state
+
+For now, the file only contains the combinatorial construction, i.e., the definition of
+`OrderedFinPartition n` and the equivalence between
+`((c : OrderedFinpartition n) × Option (Fin c.length))` and `OrderedFinpartition (n + 1)`.
+The application to the proof of the Faa di Bruno formula will be PRed in a second step.
+-/
+
+noncomputable section
+
+open Set Fin Function
+
+/-- A partition of `Fin n` into finitely many nonempty subsets, given by the increasing
```

Comment:
Is this analogous to the partitions found in #16647, applied to `finRange`?

Reply:
Yes, definitely. It looks like the order you get on sublists is even the same one as the one I need here, i.e., parts are ordered by increasing maximum. This is not so surprising because I need these to do an induction from the left (i.e., the leftmost point plays a specific role), and in your case of lists you define things by induction from the left.